### PR TITLE
Fixed security vulnerabilities/memory leaks/bugs in CryptographyHelper.cs

### PR DESCRIPTION
**Pull request by luke-park**
This PR addresses a number of the issues raised [on this StackOverflow question](https://stackoverflow.com/questions/59147167/encryption-with-same-parameters-need-different-results/59147850?noredirect=1#comment104549969_59147850) that the author more or less refused to acknowledge or fix.

I've tried to annotate the locations that needed changing.

### CODE REVIEW

File: UltimateHelperEx/CryptographyHelper.cs [hunk 1]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
-                try
+                // We can remove the try-catch and handle null strings properly.
+                if (stringToEncrypt == null)
                 {
-                    // Verify String Does Exist
-                    if (!String.IsNullOrEmpty(stringToEncrypt))
-                    {
-
-                        TripleDESCryptoServiceProvider des;
-                        MD5CryptoServiceProvider hashmd5;
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: Two of the memory leaks - these were never disposed.  ✔️ Fixed

File: UltimateHelperEx/CryptographyHelper.cs [hunk 2]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
-
-                try
+                // We can remove the try-catch and handle null strings properly.
+                if (stringToEncrypt == null)
                 {
-                    // Verify String Does Exist
-                    if (!String.IsNullOrEmpty(stringToEncrypt))
-                    {
-
-                        TripleDESCryptoServiceProvider des;
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: Insecure symmetric algorithm, changed to AES ✔️

File: UltimateHelperEx/CryptographyHelper.cs [hunk 3]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
-                        MD5CryptoServiceProvider hashmd5;
-                        byte[] pwdhash;
-                        byte[] buff;
-
-                        hashmd5 = new MD5CryptoServiceProvider();
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: MD5 was never a good KDF, even when it wasn't broken.  Fixed to use PBKDF2, which still isn't perfect, but doesn't require another dependency. ✔️

File: UltimateHelperEx/CryptographyHelper.cs [hunk 4]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
-                        pwdhash = hashmd5.ComputeHash(ASCIIEncoding.ASCII.GetBytes(password));
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: Use of ASCII instead of UTF8 - fixed ✔️

File: UltimateHelperEx/CryptographyHelper.cs [hunk 5]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
+                using (var aes = AesManaged.Create())
+                using (var pbkdf2 = new Rfc2898DeriveBytes(password, _pepper, 32767)) // MD5 is insecure as KDF, we use PBKDF2 instead.
+                using (var rng = new RNGCryptoServiceProvider())
+                {
+                    var key = pbkdf2.GetBytes(32); // Let's use AES-256.
+                    var iv = new byte[16];
+                    rng.GetBytes(iv); // We always create a new, random IV for each operation.
 
-                        // Electronic Code Book Cipher
-                        des.Mode = CipherMode.ECB; //CBC, CFB
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: ECB mode has severe vulnerabilities, fixed to use CBC with proper IV handling ✔️

File: UltimateHelperEx/CryptographyHelper.cs [hunk 6]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
 
-                        // Set Buffer To StringToEncrypt
-                        buff = ASCIIEncoding.ASCII.GetBytes(stringToEncrypt);
+                    aes.Mode = CipherMode.CBC;
+                    aes.Padding = PaddingMode.PKCS7;
+                    aes.Key = key;
+                    aes.IV = iv;
 
-                        // Get Encrypted String
-                        encryptedString = Convert.ToBase64String(des.CreateEncryptor().TransformFinalBlock(buff, 0, buff.Length));
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: Another memory leak here for the encryptor - fixed ✔️

File: UltimateHelperEx/CryptographyHelper.cs [hunk 7]
```diff
@@ -16,129 +16,65 @@ namespace DataJuggler.Core.UltimateHelper
...
+                    using (var aesTransformer = aes.CreateEncryptor())
+                    {
+                        var ciphertext = aesTransformer.TransformFinalBlock(plaintext, 0, plaintext.Length);
+                        var ivAndCiphertext = new byte[iv.Length + ciphertext.Length];
+                        Array.Copy(iv, 0, ivAndCiphertext, 0, iv.Length);
+                        Array.Copy(ciphertext, 0, ivAndCiphertext, iv.Length, ciphertext.Length);
+                        return Convert.ToBase64String(ivAndCiphertext);
                     }
                 }
-                catch
```
Comment (luke-park) [inline UltimateHelperEx/CryptographyHelper.cs]: This catch hides all exceptions - bad design.  Fixed to remove this and add null checks to input arguments ✔️

File: UltimateHelperEx/CryptographyHelper.cs (hunk 4 above)
Comment (DataJuggler) [reply UltimateHelperEx/CryptographyHelper.cs]: You updated the wrong project. You updated the Strongly Typed project, which is out of date to the other project.

**Comment by luke-park** [comment]
@DataJuggler Can you explain what you mean by "creates a different key every time you use it"?  If you provide the same password, the same key will be created - did you see the included unit test?

**Comment by DataJuggler** [comment]
I updated my Nuget package, tested, and every time the encrypted string returned is different.

It was still decryptable to the correct value, but the use case I have for this control is to take a password, encrypt it and store the encrypted value. Then if I can produce the same hash the same hash upon login I know the password input was correct.

Is there a way to have it produce the same encrypted string every time, or make it optional to be a different return value?

**Comment by luke-park** [comment]
What you're describing is completely different from an encryption algorithm - you actually want a password hashing algorithm.  Doing it the way you are describing is dangerous because it is reversible - if someone gained access to the "hash" (it's not actually a hash, it's a ciphertext), then they could decrypt it and get the plaintext password.

With a password hashing algorithm, the process that you describe stays the same, except it is *impossible* to decrypt.  Can you see the benefit of this over the original method?

**Comment by DataJuggler** [comment]
Yes a password hash is my primary use case. I thought I was creating a Hashing Algorithm, but if there right ways to do this I will implement this. 

I thought about it while in the shower just now, and my brain works better while not starring at a computer screen. The way you suggested would work if I decrypted the stored database value and compared, but as you mentioned for places where security matters this could be dangerous. 

What algorithm is best for password Hashing? 

Sorry we got off on the wrong foot on Stack Overflow, but if people who are uninformed aren't allowed to talk, they can't find out what they don't know and stay thinking they know something, like I though I knew a little about cryptography 24 hours ago. 

Corby

**Comment by luke-park** [comment]
Hi Corby,

Yeah same here, I never had any intent to have a bad interaction with you on StackOverflow.  Truth is, I'm just passionate about security and have seen what happens in the past when people use insecure crypto libraries.  It can get nasty quickly!

Regarding your question about password hashing, I would recommend Argon2 if you can find an implementation of it (and are happy to add it as a dependency) - otherwise there is bcrypt, or if you don't want to add a dependency, you can use PBKDF2, which is implemented in C# as `Rfc2898DeriveBytes`.

Luke

**Comment by DataJuggler** [comment]
I will start Googling all of this. This class wasn't designed to protect launch codes, but I appreciate all of the advice.

It is my own fault for opening my mouth, but that is how I learn.

**Comment by DataJuggler** [comment]
I just checked in a new update, and your test is back in. I added a reference to Konscious.Security.Cryptography has an implementation of Argon2. 

I read it is ok to store the salt with the password hash, and I changed the default password to NotASecret for overrides without a keycode.

Your test is back in, and most of the other changes, the only thing I did differently is you wanted to raise an error on bad input and I prefer it just to return an empty string or false on the VerifyHash method. For debugging the source code can be used, I just hate to rely on other people properly trapping an exception. 

There is also a little sample Windows Form for testing.

Is it ok if I give you credit since all the changes are per your changes / recommendations (even though I am a bad recommendation follower)? If you prefer I don't then I will just pretend I read up on all this.

Thanks for all your help.

**Comment by luke-park** [comment]
It's fine, don't worry about it - my changes are documented in the PR anyway.  Good work on the password hashing!  Happy to help.

**Pull request merged**